### ISSUE TITLE
RFC: Split generic deriving into specific cases, powered by DerivingVia.

### ISSUE THREAD

**Issue by phadej** [haskell/aeson#933]
A problem I find in improving `aeson` is dealing with generic and TH code. It's monolithic, trying to cater for all needs at once.
Issues like https://github.com/haskell/aeson/issues/824 and https://github.com/haskell/aeson/issues/827 is hard to incorporate as everything is coupled. Compiling generics code is also slow (https://github.com/haskell/aeson/issues/810) as simplifier needs to work out a lot of code.

I propose a long term strategy of splitting various kinds of encoding formats into separate newtype wrappers to be used with `DerivingVia`
- The https://github.com/haskell/aeson/issues/909 issue would be a good start, to provide the "correct" Generic deriving. (would resolve https://github.com/haskell/aeson/issues/660)
- The following step would be to provide deriving newtypes for product types. We might introduce `To/FromJSONObject` classes here.
- I propose to not touch the current deriving code for a while, until the alternatives are sufficiently expressive.
- ... but when they are, remove the current generics code and default implementations. That would be a major breaking change at that point, but on the other hand, any further changes to generics code will be much more localized.
    - That also will allow to move generics code out of `ToJSON` & `FromJSON` modules, making working on `aeson` slightly more pleasant.

"simplifying" generics story would also help to improve other aspects of aeson, e.g. https://github.com/haskell/aeson/pull/839 is somewhat stuck as getting the change through current generics code is very hard.

Note: TH code should be adopted to be configurable similarly to DerivingVia code, if people would prefer to use TH. This shouldn't be difficult.

In summary: 1. Add an alternative generics code to be used with `DerivingVia`. 2. Remove old generics code. 3. Profit.

ping @bergmark @Lysxia

**Comment by bergmark** [comment]
Sounds good!

Does DerivingVia compose well? E.g. if I want to use both `fieldLabelModifier` and `rejectUnknownFields` together, how would that look?

**Comment by phadej** [comment]
> Sounds good!
> 
> Does DerivingVia compose well? E.g. if I want to use both `fieldLabelModifier` and `rejectUnknownFields` together, how would that look?

It doesn't. Each "newtype" has to implement the features, but can use the same auxiliary code for that.

**Comment by treeowl** [comment]
> That also will allow to move generics code out of ToJSON & FromJSON modules, making working on aeson slightly more pleasant.

That approach tends to lead to orphan instances. Are those considered okay here?

**Comment by phadej** [comment]
@treeowl they won't be orphans, as in my plan there are no default definitions.

**Comment by treeowl** [comment]
But there need to be instances, right? And presumably the newtype representing the current default will be used for some of them?

**Comment by phadej** [comment]
> And presumably the newtype representing the current default will be used for some of them?

It isn't currently, I don't see why situation would dramatically change. `aeson` doesn't define instances for compound types (e.g. records with fields) where `Generic`s would be useful.

**Comment by treeowl** [comment]
Oh, didn't realize they weren't used for `Either`, tuples, `Ordering`, etc. Never mind!

**Comment by phadej** [comment]
2.0.3.0 added instance for `Generically a` which is the first `DerivingVia` newtype. More to follow.

**Comment by blackheaven** [comment]
> Sounds good!
> 
> Does DerivingVia compose well? E.g. if I want to use both `fieldLabelModifier` and `rejectUnknownFields` together, how would that look?

No, but we can have options to compose, such as in: [`deriving-aeson`](https://hackage.haskell.org/package/deriving-aeson-0.2.8/docs/Deriving-Aeson.html)

**Comment by ulidtko** [comment]
Good plan @phadej, +1 !

I just had a use-case for an extended deriving strategy... but upon reading through the `Data.Aeson.Types.FromJSON` module — I'd stopped in my tracks, realizing I'll need to duplicate / reimplement half of that module if not more.

To sketch the idea, imagine numerous serializable record-datatypes with instances of `Default`:
```haskell
data MyAirInfo = MyAirInfo
  { airTemperature :: Float
  , airPressurePa  :: !Int
  , airPlanet      :: Text
  }
  deriving Generic
  deriving ToJSON via CustomJSON (KebabCaseWithoutPrefix "air") MyAirInfo

instance Default MyAirInfo where
  def = MyAirInfo
    { airTemperature = 19.0
    , airPressurePa  = 108900
    , airPlanet      = "Earth"
    }
```
So, given that all fields have sane defaults, and can be omitted from input json...

How to write the `FromJSON` instance?

```haskell
-- either the default values are repeated in an hand-rolled parser instance:
--
instance FromJSON MyAirInfo where
  parseJSON = withObject "MyAirInfo" $ \o -> do
    airTemperature <- o .:? "temperature" .!= 19.0
    airPressurePa  <- o .:? "pressure-pa" .!= 108900
    airPlanet      <- o .:? "planet"      .!= "Earth"
    pure MyAirInfo{..}

-- or we leverage the Default instance, but repeat field names and still hand-roll:
--
instance FromJSON MyAirInfo where
  parseJSON = withObject "MyAirInfo" $ \o -> do
    airTemperature <- o .:? "temperature" .!= airTemperature def
    airPressurePa  <- o .:? "pressure-pa" .!= airPressurePa def
    airPlanet      <- o .:? "planet"      .!= airPlanet def
    pure MyAirInfo{..}

-- instead I'd want to say:
--
deriving via CustomJSONwithDefault (KebabCaseWithoutPrefix "air") MyAirInfo instance FromJSON MyAirInfo
```

— but `CustomJSONwithDefault` doesn't exist, and is currently hard to write on top of public interface of aeson.

--------------

> >    Does DerivingVia compose well? E.g. if I want to use both fieldLabelModifier and rejectUnknownFields together, how would that look?

> It doesn't. Each "newtype" has to implement the features, but can use the same auxiliary code for that.

I constructively disagree. Check out the `CustomJson` deriving-strategy here :point_right: [fumieval/deriving-aeson][]
 — just whip up a type-list of void marker types a-la `AesonOptions`, then interpret them into respective modifiers of aeson's `Options`; 200 lines, done.

And although via-strategy newtypes can also compose by nesting — I'd suggest to not restrict any particular composition style in the core aeson library. Leave it on value-level; that's sufficient. Because :point_right: [fieldstrength/aeson-deriving][] — implements similar idea, but with a different UI (or like they say in javascript circles, "DX", developer experience).

[fumieval/deriving-aeson]: https://github.com/fumieval/deriving-aeson
[fieldstrength/aeson-deriving]: https://github.com/fieldstrength/aeson-deriving

--------------

BTW, was this bug ever noticed?.. The ± infinities are flipped here.
https://github.com/haskell/aeson/blob/98875e052e9dc486ab01a1ce081cde0bbb0d7481/src/Data/Aeson/Types/FromJSON.hs#L1577-L1582

**Comment by phadej** [comment]
`CustomJSON` is not what I had in mind.

But specific encodings like `deriving (FromJSON) via Aeson.TaggedEnum [...modifiers...]`. 

And if the type cannot be encoded as `TaggedEnum` it would fail.

Modifiers may be shared (e.g. how to mangle constructor names), but that's about it.

Similarly, record encoding as object would have options, etc.

**Comment by ulidtko** [comment]
@phadej so what about supporting generic ideas like my `CustomJsonWithDefault`, do you think that'll become easy to implement? Can `FromJSON1` somehow help with that?
